### ISSUE TITLE
Create unique index on a column

### ISSUE THREAD

**Issue by remileblond67076** [gristlabs/grist-core#706]
We often need to define a unique constraint on columns, in order to prohibit the creation of duplicates in a table.

This would correspond to adding a unique index to a column in the SQLLite database.

For example : 
 ```sql
CREATE TABLE table_name(
    ...,
    column_name type UNIQUE,
    ...
);
```

Or 

```sql
CREATE UNIQUE INDEX ux_unique_name ON table_name(column_name );
```

Enabling this constraint could be set using a checkbox in the column configuration interface.

**Comment by remileblond67076** [comment]
![Grist-unique](https://github.com/gristlabs/grist-core/assets/65656372/d9d15111-d537-4c76-998b-85f2fe6bfa4f)

**Comment by fflorent** [comment]
@remileblond67076 I have started working on this issue.

Regarding the UI: 
 - Our UX expert, Aurélie, suggested that this option could be enabled for any type of columns
   - Even for lists, which would be considered as unique sets (e.g. if a cell contains `A, B`, we cannot fill `A, B` nor `B, A` in another cell of this column);
   - We are hesitant to place this option either near to the column name / description or with the type configuration;
 - I agree that the checkbox would be meaningful, however, I only see buttons to activate options on the column configuration panel;
 - But my attempts using the button is a bit awkward in this context, so despite the previous, I would probably go with a checkbox;

![2023-10-24_10-35](https://github.com/gristlabs/grist-core/assets/371705/1e3a19d2-71fa-4cce-ab3c-b0a8e08fcd21)

**Comment by fflorent** [comment]
cc'ing @kbizien

**Comment by anaisconce** [comment]
> Our UX expert, Aurélie, suggested that this option could be enabled for any type of columns

Would it be for any type of column, or any type of string column, e.g. text, choice, and choice list?

Trying to think through the practical applications of uniqueness on numeric, boolean and attachment columns. Date and datetime are also unclear to me -- maybe there's a case, but usually not. Reference and reference list also become hazier.

As for design, the table tab in the creator panel has an example of checkboxes. 

![image](https://github.com/gristlabs/grist-core/assets/69366135/1cd23e4c-f9f0-4f56-bbcf-4dc8f69c6b3c)

Re: uniqueness configuration design, if uniqueness were supported for all column types, it might make more sense to put the configuration in the column behavior section. The top section is about label, id and description. Adding uniqueness there is a bit surprising to me.
![image](https://github.com/gristlabs/grist-core/assets/69366135/d3f1015e-9525-401b-98cf-548de0d1521d)

If uniqueness is not supported for all columns, what do you think of placing uniqueness configuration after formatting and styling? This is currently what follows, and the space isn't used very well. Kevin might have better ideas.
![image](https://github.com/gristlabs/grist-core/assets/69366135/558c135c-14dd-4c06-b332-4888974d235c)

**Comment by dsagal** [comment]
Databases support unique indices on a combination of columns. For example:
```
CREATE UNIQUE INDEX idx_student_course
ON student_courses (student_id, course_id);
```
(i.e. a course enrollments table can only have a single enrollment for a particular student for a particular course).

As with SQL syntax, it makes sense to have a per-column way to say "make this column unique", but perhaps we can design it early on to support combinations of columns, so that the single-column UI is just a shortcut to creating a unique index.

Also, just wanted to add an observation: unique indices have a second benefit (besides enforcing uniqueness) of making certain operations much more efficient. So this feature will come in useful in the future when making Grist work for larger data (#43).

**Comment by paulfitz** [comment]
Small implementation note. Grist has an obscure "on-demand table" feature that uses indexes, and some of the code involved would need a small tweak to be aware of any other indexes added. Specifically:
https://github.com/gristlabs/grist-core/blob/e51c1b6b9274dbc22414824dfd823494f446b7de/app/server/lib/DocStorage.ts#L1499-L1532

So if someone makes a proof of concept and sees Grist continually deleting the index they are carefully adding, that will be why. It is no big deal, should be a small change.

**Comment by fflorent** [comment]
> Would it be for any type of column, or any type of string column, e.g. text, choice, and choice list?

For all of the types of column available, including the boolean (even if the use case is probably not interesting), integers / numbers, text, choice / choice list, ... I just wonder if that would work well with attachment columns (if that's not relevant, may it also cause performance issue? IDK).

> Trying to think through the practical applications of uniqueness on numeric, boolean and attachment columns. Date and datetime are also unclear to me -- maybe there's a case, but usually not. Reference and reference list also become hazier.

For references: say you use Grist as a task management for a team. You have 2 tables: `People` and `Tasks`. The team members can pick tasks by assigning themselves through the `Tasks` table. Also you don't want them to pick several tasks at the same time (they must unassign themselves before taking another task).

For dates, I suppose you may want to fill a calendar with one item per date.

For the other types, I admit I don't have much relevant use cases. 

> As for design, the table tab in the creator panel has an example of checkboxes.

Yes, I was thinking of the column config panel separately, not including the table config panel... Though I am probably overthinking about this, you are probably right and we should go with checkboxes.

I'll try to come back with some mockups soon so we can stimulate the discussion.

> As with SQL syntax, it makes sense to have a per-column way to say "make this column unique", but perhaps we can design it early on to support combinations of columns, so that the single-column UI is just a shortcut to creating a unique index.

Right, when you say "perhaps we can design it early on to support combinations of columns", you suggest that technically the backend supports this combination so we pave the way for future UI enhancements?

**Comment by remileblond67076** [comment]
From my point of view, it would be great to be able to take advantage of SQLLite's uniqueness constraints.
I'm not sure there's much sense in putting a uniqueness constraint on a Boolean or a list of values / references.

**Comment by emanuelegissi** [comment]
This would be great!

**Comment by lusebille** [comment]
I've made this proposition keeping the checkbox, but make it disabled when type of column is not yet selected, everything detailed here 
https://www.figma.com/design/wcpetFt6aOKzTszcvPPWLQ/%5B05%2F24%5D-Grist-Design?node-id=294-15720&t=9hvbpy8Fg7eoAurj-1


![Image](https://github.com/gristlabs/grist-core/assets/170542975/c3534334-dd7b-4413-8c50-a306636f0305)

**Comment by dsagal** [comment]
For UI, I'd like to come back to this wish: "design it early on to support combinations of columns, so that the single-column UI is just a shortcut to creating a unique index."

I propose the following:

1. Implement support for unique indexes similar to [SQLite's support](https://www.sqlite.org/lang_createindex.html).
2. Introduce "UNIQUE INDEXES" section in the creator panel's "Table" tab, under "Data" subtab.
3. Use UI similar to Sort UI (in the "Sort & Filter" subtab):

   <img width="225" alt="Screenshot 2024-07-09 at 1 08 15 AM" src="https://github.com/gristlabs/grist-core/assets/1091143/e3bdda07-7257-4933-9733-6d34cbe920dd">

   a. What's missing is that, unlike for Sort, we need to be able to add/remove indexes, give them names, and add/remove columns within each one.
   b. The options menu used for advanced sort options can be used for advanced index options, which can be added in follow-ups (e.g. case sensitivity, or treatment of empty values).
4. For the simple (and common) case of a single-column unique constraint, a checkbox in a column's creator panel as in @lusebille's design, is mostly sufficient, and should be equivalent to creating a unique index with an auto-generated name.
  a. It's a heavy operation (esp. for large tables), so a simple toggle might not be the best UX. It may be better to click a link that opens a popup, asks for confirmation, and shows a spinner.
  b. Single-column indexes should be reflected _both_ in "Table" tab - "Data" subtab; _and_ in the column creator panel; regardless of how it was created.
5. With both UI versions, if an index is created and there are duplicates in the column (or combination of columns), there should be either a question to the user, or the operation should be disallowed.

**Comment by lusebille** [comment]
I've made changes on figma according the specifications above ( I will present you next UX meeting )

**Comment by dsagal** [comment]
To clarify my request for generalization, I'd like for the feature to support multi-column indexes. For example:
- For table `Inventory`, unique index on columns `Store` and `Product`.
- For table `Events`, unique index on columns `Event_Name`, `Event_Date`, `Location`.
- For table `Orders`, index (not unique) on columns `Customer` and `Order_Date`.

A given table may have more than one index. Each index may include one or more columns. Each column may be marked "ascending" or "descending" (similar as for sorting).

Each index _probably_ should have a name. Not sure since names could be auto-generated, but a good name could make it easier. Even an auto-generated name needs to be visible, as it is sometimes needed for some database operations.

Each index may specify "UNIQUE" or not. A "unique" index should prevent duplicate entries from being stored. A non-unique index has different purposes (mainly, efficiency of certain queries and operations), but is still valuable.

**Comment by emanuelegissi** [comment]
IMHO, what @dsagal wrote is of paramount importance.

**Comment by paulfitz** [comment]
> Each column may be marked "ascending" or "descending" (similar as for sorting).

This is useful, but am I seeing a significant increase in scope from the original goal of specifying a uniqueness constraint?

> Each index may specify "UNIQUE" or not.

@dsagal does this extend the scope of the eventual implementation to indexing in general?

**Comment by dsagal** [comment]
> > Each index may specify "UNIQUE" or not.
> 
> @dsagal does this extend the scope of the eventual implementation to indexing in general?

Yes, this is a generalization of the original request, and is intended to support indexing in general. The motivation is that we think we'll want general indexing for cases when it matters for performance, and it would be more efficient for us to design and build that, with unique indexes as a special case of the more general indexing.

**Comment by ludo427** [comment]
Hi everyone, any updates on this issue ?

**Comment by toihr** [comment]
I am also interested if this is still worked on? As to my understanding the underlying data is stored in a SQlite file anyways, and does sqlite not already support UNIQUE?
